### PR TITLE
Add gigl.Module with RetrievalLoss and LinkPredictionGNN

### PR DESCRIPTION
**Pull request by mkolodner-sc**
**Scope of work done**

Currently, it is a bit awkward to use the existing LinkPredictionGNN and RetrievalLoss due to assumptions we made as part of offline SGS GiGL. For example, we need to provide a `output_node_types` to the forward pass of the model even if we are in the homogeneous setting, and index into a dictionary as a result as well when getting the output from the forward pass. Additionally, we need to populate an opinionated `BatchCombinedScores` dataclass in order to calculate RetrievalLoss. This is not PyG-conformant. 

This PR moves these two fields to their own files in a new `gigl.module` folder and removes these awkward assumptions from these classes. 

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** NO

### CODE REVIEW

File: python/gigl/module/loss.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            repeated_query_ids (torch.Tensor): Repeated query user IDs. Tensor shape: [num_positives]
+            candidate_sampling_probability (Optional[torch.Tensor]): Optional tensor of candidate sampling probabilities.
+                When given will be used to correct the logits to reflect the sampling probability of negative candidates.
+                Tensor shape: [num_positives + num_hard_negatives + num_random_negatives]
+        """
+        if repeated_query_ids.numel():
+            loss = self._calculate_batch_retrieval_loss(
+                scores=repeated_candidate_scores,
+                candidate_sampling_probability=candidate_sampling_probability,
+                query_ids=repeated_query_ids,
```
Comment (yliu2-sc) [inline python/gigl/module/loss.py:186]: would the behavior be the same if you just feed `None` when there are no elements in `repeated_query_ids` and let `_calculate_batch_retrieval_loss` handle it?

**Review by yliu2-sc — APPROVED** [review]

File: python/gigl/module/models.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+from gigl.src.common.models.pyg.link_prediction import LinkPredictionDecoder
+from gigl.src.common.types.graph_data import NodeType
+
+
+class LinkPredictionGNN(nn.Module):
+    """
+    Link Prediction GNN model for both homogeneous and heterogeneous use cases
+    Args:
+        encoder (nn.Module): Either BasicGNN or Heterogeneous GNN for generating embeddings
```
Comment (svij-sc) [inline python/gigl/module/models.py:15]: If these (BasicGNN, Heterogeneous GNN) are GiGL classes, lets decorate them so sphinx doc can link them.
See example here: https://snapchat.github.io/GiGL/_modules/gigl/distributed/utils/networking.html#get_internal_ip_from_master_node

Using the following
```
    using :py:obj:`gigl.distributed.dataset_factory`
```

visualizes as:
https://snapchat.github.io/GiGL/docs/api/gigl/distributed/utils/networking/index.html#gigl.distributed.utils.networking.get_internal_ip_from_master_node

Screenshot 2025-07-01 at 11 04 27 PM

File: python/gigl/src/common/models/layers/loss.py [hunk 3]
```diff
@@ -205,6 +208,10 @@ def __init__(
                 f"The temperature is expected to be greater than 1e-12, however you provided {self._temperature}"
             )
         self._remove_accidental_hits = remove_accidental_hits
+        logger.warning(
+            "Calculating retrieval loss with the class at gigl.src.common.models.layers.loss.RetrievalLoss is deprecated and will be removed in a future release. "
+            "Please use the `gigl.module.loss.RetrievalLoss` class instead."
```
Comment (svij-sc) [inline python/gigl/src/common/models/layers/loss.py:213]: Same comment as  above.

Lets, make sure these class / object paths get dynamically linked them using following syntax :py:obj:`gigl.module.loss.RetrievalLoss`. This will help us create better docs.

cc @kmonte , and @yliu2-sc as well for visibility to help improve our documentation

**Review by svij-sc — COMMENTED** [review]
Just some nits for the future.
